### PR TITLE
Fix for incorrect variable name handle

### LINKED ISSUE #402: Use new variable searching mechanism to correctly rename local properties
In simple cases like
```
val foo = Foo
use(foo)
```
we can rename `foo` without breaking code

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
   Added fix for incorrect variable handle and tests
  closes #402

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt [hunk 1]
```diff
@@ -162,6 +166,11 @@ class IdentifierNaming(private val configRules: List<RulesConfig>) : Rule("ident
                     // variable name should be in camel case. The only exception is a list of industry standard variables like i, j, k.
                     VARIABLE_NAME_INCORRECT_FORMAT.warnAndFix(configRules, emitWarn, isFixMode, variableName.text, variableName.startOffset, node) {
                         // FixMe: cover fixes with tests
+                        variableName
+                            .parent({it.elementType == FILE})
+                            ?.findAllVariablesWithUsages { it.name == variableName.text && (it.isLocal || it.isPrivate()) }
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt:171]: I think, the `variableName` node should be local or private, what's the point in checking it's usages?

File: diktat-rules/src/test/resources/test/paragraph1/naming/identifiers/VariableNamingTest.kt [hunk 2]
```diff
@@ -10,3 +10,18 @@ private val loWer_VAlue = ""
 private val lower = ""
 private val ValN_x256 = ""
 private val VoiceIP_port = ""
+
+class A {
```
Comment (petertrr) [inline diktat-rules/src/test/resources/test/paragraph1/naming/identifiers/VariableNamingTest.kt:14]: We need more tests: how are local variables renamed, how are variables from companion object renamed when used in class (does `findAllVariablesWithUsages` find even them?).

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt [hunk 3]
```diff
@@ -134,10 +140,11 @@ class IdentifierNaming(private val configRules: List<RulesConfig>) : Rule("ident
     /**
      * all checks for case and naming for vals/vars/constants
      */
-    @Suppress("SAY_NO_TO_VAR")
+    @Suppress("SAY_NO_TO_VAR", "TOO_LONG_FUNCTION", "ComplexMethod")
     private fun checkVariableName(node: ASTNode): List<ASTNode> {
         // special case for Destructuring declarations that can be treated as parameters in lambda:
         var namesOfVariables = extractVariableIdentifiers(node)
+        val isFix = isFixMode && if (node.elementType == PROPERTY) (node.psi as KtProperty).run { isLocal || isPrivate() } else true
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt:147]:
```suggestion
        // only local private properties will be autofix in order not to break code if there are usages in other files
        val isFix = isFixMode && (node.psi as? KtProperty)?.run { isLocal || isPrivate() } ?: true
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt [hunk 4]
```diff
@@ -154,14 +161,19 @@ class IdentifierNaming(private val configRules: List<RulesConfig>) : Rule("ident
...
                         }
                     }
                 } else if (variableName.text != "_" && !variableName.text.isLowerCamelCase()) {
                     // variable name should be in camel case. The only exception is a list of industry standard variables like i, j, k.
-                    VARIABLE_NAME_INCORRECT_FORMAT.warnAndFix(configRules, emitWarn, isFixMode, variableName.text, variableName.startOffset, node) {
+                    VARIABLE_NAME_INCORRECT_FORMAT.warnAndFix(configRules, emitWarn, isFix, variableName.text, variableName.startOffset, node) {
                         // FixMe: cover fixes with tests
+                        variableName
+                            .parent({it.elementType == FILE})
+                            ?.findAllVariablesWithUsages { it.name == variableName.text && (it.isLocal || it.isPrivate()) }
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt:174]: I still don't understand why check for local and private is needed here. Your rule should be able to rename variable in cases like
```kotlin
val foo = 42
bar(foo)
```
Will it handle it correctly?

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt (hunk 4 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt:174]: Yes, I added to the test

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt [hunk 5]
```diff
@@ -134,10 +140,12 @@ class IdentifierNaming(private val configRules: List<RulesConfig>) : Rule("ident
     /**
      * all checks for case and naming for vals/vars/constants
      */
-    @Suppress("SAY_NO_TO_VAR")
+    @Suppress("SAY_NO_TO_VAR", "TOO_LONG_FUNCTION", "ComplexMethod")
     private fun checkVariableName(node: ASTNode): List<ASTNode> {
         // special case for Destructuring declarations that can be treated as parameters in lambda:
         var namesOfVariables = extractVariableIdentifiers(node)
+        // only local private properties will be autofix in order not to break code if there are usages in other files
+        val isFix = isFixMode && if (node.elementType == PROPERTY) (node.psi as KtProperty).run { isLocal || isPrivate() } else true
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt:148]: Also, `namesOfVariables` extracted from `node` is a collection, while you only check autofix condition for the `node` itself.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt [hunk 6]
```diff
@@ -154,14 +162,19 @@ class IdentifierNaming(private val configRules: List<RulesConfig>) : Rule("ident
...
                 } else if (variableName.text != "_" && !variableName.text.isLowerCamelCase()) {
                     // variable name should be in camel case. The only exception is a list of industry standard variables like i, j, k.
-                    VARIABLE_NAME_INCORRECT_FORMAT.warnAndFix(configRules, emitWarn, isFixMode, variableName.text, variableName.startOffset, node) {
+                    VARIABLE_NAME_INCORRECT_FORMAT.warnAndFix(configRules, emitWarn, isFix, variableName.text, variableName.startOffset, node) {
                         // FixMe: cover fixes with tests
+                        variableName
+                            .parent({it.elementType == FILE})
+                            ?.findAllVariablesWithUsages { it.name == variableName.text }
+                            ?.flatMap { it.value.toList() }
+                            ?.forEach { (it.node.firstChildNode as LeafPsiElement).replaceWithText(variableName.text.toLowerCamelCase())}
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt:177]: This `variableName.text.toLowerCamelCase()` should be extracted to a separate variable like `correctVariableName`.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt [hunk 7]
```diff
@@ -134,10 +140,13 @@ class IdentifierNaming(private val configRules: List<RulesConfig>) : Rule("ident
...
      * all checks for case and naming for vals/vars/constants
      */
-    @Suppress("SAY_NO_TO_VAR")
+    @Suppress("SAY_NO_TO_VAR", "TOO_LONG_FUNCTION", "ComplexMethod")
     private fun checkVariableName(node: ASTNode): List<ASTNode> {
         // special case for Destructuring declarations that can be treated as parameters in lambda:
         var namesOfVariables = extractVariableIdentifiers(node)
+        // only local private properties will be autofix in order not to break code if there are usages in other files
+        val isFix = isFixMode && if (node.elementType == PROPERTY) (node.psi as KtProperty).run { isLocal || isPrivate() } else true
+        // Destructuring declarations are only allowed for local variables/values
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/IdentifierNaming.kt:149]:
```suggestion
        // Only local private properties will be autofix in order not to break code if there are usages in other files.
        // Destructuring declarations are only allowed for local variables/values, so we don't need to calculate `isFix` for every node in `namesOfVariables`
        val isFix = isFixMode && if (node.elementType == PROPERTY) (node.psi as KtProperty).run { isLocal || isPrivate() } else true
```

**Review by petertrr — APPROVED** [review]

**Pull request merged**